ipsec: add support for using remote PodCIDR entries

julianwiedmann opened the pull request:
Update the encryption policy for IPsec, so that it can take advantage of a PodCIDR entry for the destination endpoint.

Add a first set of encryption-related tests that cover scenarios with terminating endpoints.

smagnani96 commented on bpf/tests/pktgen.h:
```diff
@@ -67,6 +67,13 @@ volatile const __u8 mac_zero[] = mac_zero_addr;
 #define v4_pod_two	IPV4(192, 168, 0, 2)
 #define v4_pod_three	IPV4(192, 168, 0, 3)
 
+#define v4_pod_one_on_node_two		IPV4(192, 168, 1, 1)
```
After https://github.com/cilium/cilium/pull/40294, we haven't come up yet with a good practise to keep this file in sync with the newly introduced https://github.com/cilium/cilium/blob/main/bpf/tests/scapy/pkt_defs.py.
Maybe it is worth adding these bits also there?

julianwiedmann replied:
oh snap ... yes, we'll need to see about porting the existing tests over then I guess. Will add it for now.

smagnani96 commented on bpf/tests/encrypt_host.h:
```diff
@@ -0,0 +1,242 @@
+
+	pktgen__init(&builder, ctx);
+
+	l3 = pktgen__push_ipv4_packet(&builder, (__u8 *)mac_one, (__u8 *)mac_two,
```
Just fyi, here it doesn't seem needed (don't need to manipulate the packet too much), but in  https://github.com/cilium/cilium/pull/40294 we added a new way to define packets.

smagnani96 commented on bpf/tests/encrypt_host.h:
```diff
@@ -0,0 +1,242 @@
+	test_finish();
+}
+
+/* Finally test without the mark, but with the endpoint's ipcache entry: */
```
Maybe I'm tripping, but is it worth adding the case in which we don't have endpoint ipcache entry but still local CIDR entry in the ipcache? 🤔

julianwiedmann replied:
I'd need to double-check, but IIRC we currently don't push an ipcache entry for the local PodCIDR.

smagnani96 replied:
Right... I did a quick test and the only CIDRs I see from a node are the remote ones 😞 

Don't know if it would make sense to add (i.e., if there's a time window where we would already have CIDRs but not endpoint info locally).

julianwiedmann replied:
I also see the value of having the local PodCIDR in the ipcache entry, yep. It needs some architecture-level thinking though:
* for outbound traffic, we should ideally rely on the source identity in the packet mark. And **not** need an ipcache lookup, or endpoint map lookup.
* for inbound traffic, it would be cool to drop any un-deliverable packet that's intended for the local PodCIDR. And not exit via the default route.

smagnani96 replied:
> * for inbound traffic ...

You mean adjusting [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_host.c#L784-L788) like:

```c
	/* Below remainder is only relevant when traffic is pushed via cilium_host.
	 * For traffic coming from external, we're done here.
	 */
	if (!from_host) {
		if (ip4->daddr MATCH localPodCIDR) {
			return DROP_EP_NOT_READY; // or other drop reason, eg DROP_UNROUTABLE
		}
		return CTX_ACT_OK;
	}
	...
```

julianwiedmann replied:
yep! And similar in the other inbound programs (from-overlay, from-wireguard, from-container).

pippolo84 replied:
> I'd need to double-check, but IIRC we currently don't push an ipcache entry for the local PodCIDR.

That's correct, we do not push the local node podCIDR at the moment:

https://github.com/cilium/cilium/blob/6263cd269e2ab877bbe7311794e7aa78bb8d47ba/pkg/node/manager/manager.go#L775-L792

AFAIR the only reason not to do it was that it was simply out of scope for https://github.com/cilium/cilium/pull/38483

smagnani96 reviewed (commented):
> @smagnani96 wdyt, does this make sense so far? Then I'd go and add IPv6 support as well, along with polishing it a bit.

Yep, just left few nits, and in case a potential idea for a new test.
Replicating the same for IPv6 too LGTM.

julianwiedmann committed:
bpf: ipsec: remove special-case for IPPROTO_ESP in encryption policy

We already check whether the packet has been encrypted by Cilium (by
looking at the skb->mark), so there's no more risk of accidental double
encryption.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: ipsec: relax encryption policy for remote endpoint

When the encryption policy was still enforced in from-container, we applied
three criteria:
1. the source is a pod (naturally, as the traffic was observed by bpf_lxc),
2. the destination has a valid `tunnel_endpoint,
3. the effective SPI is non-zero.

Slim down our new encryption policy, so that it only requires a valid
`tunnel_endpoint` and `key` for the remote endpoint. This is sufficient
to take advantage of the new PodCIDR entries in the ipcache (which have a
WORLD_ID identity).

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: test: cover terminating endpoints with IPsec / Wireguard

Walk through the different scenarios we can encounter when the local /
remote endpoint is being torn down, and their ipcache entry is not
available.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

smagnani96 reviewed (approved):
LGTM, thanks!
(Tests for IPv6 TBD, can go separately)

ldelossa commented on bpf/lib/encrypt.h:
```diff
@@ -182,12 +182,9 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
-ipsec_redirect_sec_id_ok(__u32 src_sec_id, __u32 dst_sec_id, int ip_proto) {
-	if (ip_proto == IPPROTO_ESP)
-		return 0;
+ipsec_redirect_sec_id_ok(__u32 src_sec_id, __u32 dst_sec_id) {
```
This change is okay, as long as no OTHER ESP related packets are coming through the datapath.

I think this is kinda an assumed situation if IPsec is enabled in Cilium, but just calling this out, since we are now in a pretty generic spot (netdev egress).

julianwiedmann replied:
A pod will not be able to send ESP traffic, or receive it (since we can't conn-track it). So I think we're safe, if the host is sending its own ESP traffic then we won't encrypt it with our current policy.

ldelossa replied:
> host is sending its own ESP traffic then we won't encrypt it with our current policy

As long as its always given a sec ID. I guess any other case would be a bug?

julianwiedmann replied:
yep, we rely on genuine host-originating traffic to be marked by our iptables rule with `HOST_ID`.

ldelossa commented on bpf/lib/encrypt.h:
```diff
@@ -291,10 +281,10 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto,
 	}
 
-	if (!dst || !dst->flag_has_tunnel_ep)
+	if (!dst || !dst->flag_has_tunnel_ep || !dst->key)
```
nit, can this go into the `ipsec_redirect_sec_id_ok` function, so policy is done all in one place? I.e. pass `dst` into `ipsec_redirect_sec_id_ok`. IMO, this would help to understand in one place the policy.

julianwiedmann replied:
If we then also rename the function, sure ;). But tbh I actually like the idea of evaluating encryption-eligibility separately, once for source and once for destination.

ldelossa replied:
Changing name sounds good. IMO, its nicer to have all the policy stuff in one function, so its clear exactly where the policy occurs. But I don't feel too strongly. Hence the nit.

julianwiedmann replied:
ack. Let's see how well-aligned IPsec and Wireguard are by now, ideally they'd be using the same helpers ...

ldelossa commented on bpf/lib/encrypt.h:
```diff
@@ -177,24 +177,14 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
-		return 0;
 	if (!identity_is_cluster(src_sec_id))
 		return 0;
-	if (identity_is_remote_node(dst_sec_id))
```
Checking for remote host is done via IPCache now as well right? Key will be 0 and tunnel_endpoint will be 0.

julianwiedmann replied:
yep, I consider the `key` as the authoritative statement whether the destination wants encryption. 

A missing `tunnel_endpoint` might then even be signal that we should drop with `NO_TUNNEL_ENDPOINT`, since we're supposed to encrypt but can't. Let's see.

ldelossa reviewed (approved):
LGTM.
